src/lib/OS/SimOS.cc

```diff
@@ -0,0 +1,228 @@
+    : executablePath_(executablePath),
+      executableArgs_(executableArgs),
+      memory_(mem),
+      syscallHandler_(std::make_shared<SyscallHandler>(processes_)) {
+  createInitialProcess();
+
+  // Create the Special Files directory if indicated to do so in Config file
+  if (Config::get()["CPU-Info"]["Generate-Special-Dir"].as<bool>() == true)
+    createSpecialFileDirectory();
+}
+
+void SimOS::tick() {
+  // Check if simulation halted
+  if (halted_) {
+    return;
+  }
+
+  if (processes_.size() == 0) {
+    halted_ = true;
+    return;
+  }
+
+  /** Scheduling behaviour :
+   * 1. All processes start in a waiting state, and are placed in the
+   * 'waitingProc' queue.
+   *
+   * 2. All cores are looped over, checking their status. If there is a process
+   * at the front of the waitingProc queue then a currently executing core will
+   * be tested to see if it should perform a context switch to the waiting
+   * process. If the core has been executing for longer than execTicks then an
+   * interupt signal is sent to test if a context switch can be made.
+   *
+   * 3. If the interrupt signal is successful, then the head of waitingProcs
+   * queue is put into scheduled queue.
+   *
+   * 4. In order to schedule a process onto a core, a core must be in an Idle
+   * state. An Idle core can only be scheduled a process from the scheduledProcs
+   * queue, i.e. a process which has previously triggered a successful interupt
+   * signal.
+   *
+   * 5. When a process is successfully scheduled, it is removed from the
+   * scheduledProcs queue.
+   *
+   * 6. When a process is de-scheduled from a core it goes back into a waiting
+   * state and is pushed to the back of the waitingProcs queue.
+   *
+   * By using two queues, we ensure that a single process will not interupt
+   * multiple cores. This is due to :
+   *    a. Only a process in the waitingProc queue can cause an interupt
+   *    b. On a successful interupt, the process that caused it is moved to the
+   * scheduledProc queue
+   *    c. Only processes in the scheduledProc queue can be scheduled onto a
+   * core
+   *
+   * If not for this process, then a waitingProc could tell more executing cores
+   * to context switch given a core is likely to be in a switching state (post
+   * successful interupt, pre Idle state) for multiple cycles.
+   * */
+
+  // Loop over all cores, apply correct behaviour based on state
+  for (auto core : cores_) {
+    switch (core->getStatus()) {
+      case CoreStatus::halted: {
+        // Core has had fatal fault. Change SimOS status and return
+        halted_ = true;
+        return;
+      }
+      case CoreStatus::idle: {
+        // Core is idle, schedule head of scheduledProc queue
+        if (!scheduledProcs_.empty()) {
+          // Get context of process that was executing on core before interrupt
+          // was signalled
+          OS::cpuContext prevContext = core->getCurrentContext();
```

Comment:
the function get the current context but then puts it into a variable representing a previous context. Think it would be best to have a consistent scope here

Reply:
+1

Reply:
Changed inline with getCurrentContext